Add `gr_mat_companion` to construct companion matrices

rburing opened the pull request:
I would like to apply this to Ore polynomials, but it works for generic polynomials so here it is (with tests).

fredrik-johansson commented:
Two small optimizations: you could apply ``gr_mat_one`` / ``gr_mat_set_scalar`` to a window matrix and then use ``_gr_vec_neg`` / ``_gr_vec_mul_scalar`` to write the poly coefficients. Also, it would be handy to have underscore versions that take the array of coefficients as input.

rburing commented:
I'll add the underscore versions in this PR. Since the optimizations also apply to the three existing `companion.c` (for `ca_mat`, `acb_mat` and `arb_mat`), can I do it in a separate PR? Also, if I understood correctly then doing it that way would still require to initialize the first column (except the bottom entry), and we don't have a strided setter yet, though it could also be achieved with a window matrix and `gr_mat_zero`.

fredrik-johansson commented:
You might also want to put error handling in the non-underscore methods (return ``GR_DOMAIN`` if the matrix shape doesn't match the polynomial).

fredrik-johansson commented:
> Since the optimizations also apply to the three existing `companion.c` (for `ca_mat`, `acb_mat` and `arb_mat`), can I do it in a separate PR?

Best would be to turn these into wrappers of ``gr_mat_companion`` to reduce code size, though this is slightly nontrivial since one wants error handling: ``arb_mat`` and ``acb_mat`` would need to handle a ``!= GR_SUCCESS`` result and then call ``arb_mat_indeterminate`` / ``acb_mat_indeterminate``, and ``ca_mat_companion`` needs to return 0 or 1 according to the present interface.

rburing commented:
Thanks, I processed all the suggestions. I removed the unused `len` parameter from the underscore methods. For `arb_mat` and `acb_mat` I called `arb_mat_indeterminate` / `acb_mat_indeterminate` and removed the `flint_throw`, assuming that's what you meant.

I tried to use `_gr_mat_companion` in `ca_mat_companion` with `_gr_ctx_init_ca_from_ref`, but I didn't manage to make it work. The underscore method `_ca_mat_companion` (undocumented) also has a different signature.
